components/module-components/blog/TagInput.tsx

```diff
@@ -0,0 +1,111 @@
+/* eslint-disable jsx-a11y/click-events-have-key-events */
+/* eslint-disable jsx-a11y/no-static-element-interactions */
+
+import React, { ChangeEvent, useEffect, useState } from 'react';
+import { toast } from 'react-toastify';
+import styles from './TagInput.module.scss';
+import useGetTags from '../../../apis/queries/tags/getTags';
+import useDebounce from '../../../utils/hooks/useDebounce';
+import useAddNewTag from '../../../apis/mutations/tags/useAddNewTag';
+
+export interface TagObject {
+  id: number;
+  name: string;
+  created_at: string;
+  updated_at: string;
+}
+
+interface TagInputProps {
+  selectedTags: TagObject[];
+  onTagsChange: (tags: TagObject[]) => void;
+}
+
+const TagInput = ({ selectedTags, onTagsChange }: TagInputProps) => {
+  const [searchTag, setSearchTag] = useState<string>('');
+  const [filteredTags, setFilteredTags] = useState<TagObject[]>([]);
+
+  const debouncedSearchTerm = useDebounce(searchTag, 300);
+  const { data } = useGetTags(debouncedSearchTerm);
+
+  const addNewTagMutation = useAddNewTag();
+
+  const addNewTag = async (tagName: string) => {
+    const newTag = await addNewTagMutation.mutateAsync({ tagName });
+    onTagsChange([...selectedTags, newTag.data as TagObject]);
+
+    setSearchTag('');
+    setFilteredTags([]);
+  };
+
+  useEffect(() => {
+    if (data?.data) {
```

Comment:
We should avoid coupling between the API layer and our components. We shouldn't transform the response data inside the components. (data.data)
This should be done inside the useGetTags query.

Reply:
Thank you for your feedback!

I want to make sure I fully understand your suggestion. Are you recommending that:

*   The useGetTags hook should return only the data field (omitting fields like "links" and "meta") to simplify the response? (this kind of seems unnecessary to me... what if we need to use the other fields somewhere else?)
*   Or, that the transformation logic (e.g., converting name to lowercase) should be moved to the useGetTags hook itself, so the component receives already-processed data?

Let me know so I can better address the issue.

Reply:
@petark7 
I’m recommending the second option: that transformation logic (e.g., processing data.data, converting names to lowercase) should be moved into the useGetTags hook. This will make sure that the component receives data that’s already in the desired format, we will have better separation of concerns and we will reduce the chance of repetitive logic in different components.

To clarify, I’m not suggesting the hook should not return fields like links or meta, as they might be useful elsewhere.
(you can use the select option in useQuery to return all of them).

The goal is just to avoid components having to directly manipulate the raw API response (like data.data), which keeps them simpler and focused on rendering.

Reply:
Since this is just a suggestion, I am also approving the PR.

Reply:
Since this is just a suggestion, I am also approving the PR.